File: mpsc-event-pipeline-kata/cpp/bench/correctness.cpp
// Phase 1 — correctness under stress, and producer-slot churn.
//
// These are the assertions that make the rest of the numbers worth reading: if
// ordering or accounting is wrong, a fast pipeline is just a fast way to be
// wrong.

#include "harness.hpp"
#include "phases.hpp"

#include "event_pipeline.hpp"
#include "sink.hpp"

#include <array>
#include <atomic>
#include <cstdio>
#include <string>
#include <thread>
#include <vector>

namespace weir::bench {

void test_correctness(ScanPolicy scan) {
    const char* name = (scan == ScanPolicy::Bitmap) ? "bitmap" : "full-scan";
    constexpr std::uint32_t kProducers = 16;
    constexpr std::uint32_t kPerProducer = 200000;

    PipelineConfig cfg;
    cfg.scan = scan;
    cfg.full = FullPolicy::SpinThenDrop;
    EventPipeline pipe(cfg);

    // Per-producer ordering check, run on the consumer thread.
    std::array<std::uint64_t, kMaxProducers> last{};
    std::array<bool, kMaxProducers> seen{};
    std::atomic<std::uint64_t> order_violations{0};
    std::atomic<std::uint64_t> received{0};
    pipe.on_event_ = [&](const Event& e) {
        const std::uint32_t i = e.producer_id;
        // Gaps are legal (a drop), going backwards is not.
        if (seen[i] && e.seq <= last[i]) order_violations.fetch_add(1, std::memory_order_relaxed);
        seen[i] = true;
        last[i] = e.seq;
        received.fetch_add(1, std::memory_order_relaxed);
    };

    CountingSink sink;
    pipe.start(sink);

    std::vector<std::thread> producers;
    for (std::uint32_t p = 0; p < kProducers; ++p) {
        producers.emplace_back([&pipe] {
            auto h = pipe.register_producer();
            if (!h.valid()) return;
            for (std::uint32_t s = 1; s <= kPerProducer; ++s)
                h.push(make_event(h.id(), s, now_ns()));
        });
    }
    for (auto& t : producers) t.join();
    pipe.stop(StopMode::Drain);

    const auto st = pipe.stats();
    std::printf("  [%s] pushed=%llu dropped=%llu received=%llu order_violations=%llu\n",
                name,
                static_cast<unsigned long long>(st.pushed),
                static_cast<unsigned long long>(st.dropped),
                static_cast<unsigned long long>(received.load()),
                static_cast<unsigned long long>(order_violations.load()));

    check(order_violations.load() == 0, std::string(name) + ": per-producer order preserved");
    check(st.pushed + st.dropped == std::uint64_t(kProducers) * kPerProducer,
          std::string(name) + ": every push accounted for (accepted + dropped)");
    check(received.load() == st.pushed,
          std::string(name) + ": drain shutdown lost nothing (received == pushed)");
    check(pipe.drain_completed(), std::string(name) + ": drain finished within deadline");
    check(st.high_water <= kRingCapacity, std::string(name) + ": ring depth stayed bounded");
    check(sink.batches() > 0 && sink.bytes() == received.load() * sizeof(Event),
          std::string(name) + ": sink byte accounting matches");
}

void test_producer_lifetime() {
    // Producers that come and go: slots must be recycled, and a ring must never
    // be reused before the consumer has drained the retired producer's events.
    PipelineConfig cfg;
    EventPipeline pipe(cfg);
    CountingSink sink;
    std::atomic<std::uint64_t> received{0};
    pipe.on_event_ = [&](const Event&) { received.fetch_add(1, std::memory_order_relaxed); };
    pipe.start(sink);

    constexpr int kWaves = 40, kThreads = 8, kPer = 500;
    for (int w = 0; w < kWaves; ++w) {
        std::vector<std::thread> ts;
        for (int t = 0; t < kThreads; ++t) {
            ts.emplace_back([&] {
                auto h = pipe.register_producer();
                if (!h.valid()) return;
                for (int s = 1; s <= kPer; ++s) h.push(make_event(h.id(), std::uint32_t(s), 0));
            });   // handle dtor retires the slot here, while the consumer may still be draining
        }
        for (auto& t : ts) t.join();
    }
    pipe.stop(StopMode::Drain);

    const auto st = pipe.stats();
    std::printf("  churn: %d waves x %d producers, pushed=%llu received=%llu\n",
                kWaves, kThreads,
                static_cast<unsigned long long>(st.pushed),
                static_cast<unsigned long long>(received.load()));
    check(received.load() == st.pushed, "lifetime: retired producers' events all drained");
    check(st.dropped == 0, "lifetime: slot recycling did not lose capacity");
}

}  // namespace weir::bench

File: mpsc-event-pipeline-kata/go/internal/phases/correctness.go
package phases

import (
	"fmt"
	"sync"
	"sync/atomic"

	"weir/internal/verify"
	"weir/internal/weir"
)

// Correctness is phase 1: the same four invariants the C++ asserts, under both
// scan policies, with the same sentence-shaped labels.
//
// The ordering assertion allows gaps but only where a drop was counted. A
// producer whose push failed has a hole in its sequence by design; the claim is
// that events never arrive out of order, not that none are lost.
func Correctness(short bool) {
	fmt.Println("== 1. correctness under stress ==")

	const producers = 16
	perProducer := 200_000
	if short {
		perProducer = 20_000
	}

	for _, scan := range []weir.ScanPolicy{weir.FullScan, weir.Bitmap} {
		clock := weir.NewClock()
		cfg := weir.DefaultConfig()
		cfg.Scan = scan
		p := weir.New(cfg, clock)

		// The hook runs only on the consumer goroutine, so this needs no
		// synchronization: single writer by construction.
		lastSeq := make([]int64, weir.MaxProducers)
		for i := range lastSeq {
			lastSeq[i] = -1
		}
		var violations, received int64
		p.SetEventHook(func(e *weir.Event) {
			received++
			if int64(e.Seq) <= lastSeq[e.ProducerID] {
				violations++
			}
			lastSeq[e.ProducerID] = int64(e.Seq)
		})

		p.Start(weir.NewCountingSink())
		var wg sync.WaitGroup
		for i := 0; i < producers; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				h, ok := p.RegisterProducer()
				if !ok {
					return
				}
				defer h.Close()
				for s := 0; s < perProducer; s++ {
					h.Push(weir.Event{ProducerID: h.ID(), Seq: uint32(s), StampNs: clock.Nanos()})
				}
			}()
		}
		wg.Wait()
		p.Stop(weir.Drain)

		st := p.Stats()
		offered := uint64(producers * perProducer)
		name := scan.String()

		fmt.Printf("  [%s] pushed=%d dropped=%d received=%d order_violations=%d\n",
			name, st.Pushed, st.Dropped, received, violations)

		verify.Check(violations == 0, name+": per-producer order preserved")
		verify.Check(st.Pushed+st.Dropped == offered,
			name+": every push accounted for (accepted + dropped)")
		verify.Check(uint64(received) == st.Pushed,
			name+": drain shutdown lost nothing (received == pushed)")
		verify.Check(p.DrainCompleted(), name+": drain finished within deadline")
		verify.Check(st.HighWater <= weir.RingCapacity, name+": ring depth stayed bounded")
	}

	producerLifetime(short)
}

// producerLifetime is phase 1b: slots recycled through many waves, asserting a
// retired producer's events are never lost to slot reuse.
//
// The counter fold at reclaim is the subtle part, and it caused a real bug in
// the C++ version: per-slot counters are reset at registration, so without
// folding a departing producer's totals into lifetime counters first, a recycled
// slot erases its predecessor's history and the accounting silently stops
// adding up.
func producerLifetime(short bool) {
	waves, threads, per := 40, 8, 500
	if short {
		waves = 8
	}

	clock := weir.NewClock()
	p := weir.New(weir.DefaultConfig(), clock)
	var received atomic.Int64
	p.SetEventHook(func(*weir.Event) { received.Add(1) })
	p.Start(weir.NewCountingSink())

	for w := 0; w < waves; w++ {
		var wg sync.WaitGroup
		for i := 0; i < threads; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				h, ok := p.RegisterProducer()
				if !ok {
					return
				}
				for s := 0; s < per; s++ {
					h.Push(weir.Event{ProducerID: h.ID(), Seq: uint32(s)})
				}
				h.Close() // retire while the consumer may be mid-drain
			}()
		}
		wg.Wait()
	}
	p.Stop(weir.Drain)

	st := p.Stats()
	want := uint64(waves * threads * per)
	fmt.Printf("  churn: %d waves x %d producers, pushed=%d received=%d\n",
		waves, threads, st.Pushed, received.Load())
	verify.Check(st.Pushed+st.Dropped == want,
		"churn: accounting survived slot recycling")
	verify.Check(uint64(received.Load()) == st.Pushed,
		"churn: every event from a retired producer was drained")
}
